### PR TITLE
DEP: Deprecate conversion of out-of-bound Python integers

### PR DESCRIPTION
**Pull request by seberg**
Any conversion from a Python integer (or subclass) that is stored
into a NumPy dtype but does not fit should raise an error in the future.
Note, that casts between NumPy types (or assignments of them) are
explicitly not affected by this.

---

A few important notes:
* One reason here was to simplify gh-21875, since it requires one special path less.
* The test changes are a bit tedious but seem mostly OK (sometimes you have to do a cast dance).
* The changes to the ufunc identity I have no sqirms about (gh-20970 should allow to fix it up)
* The changes to MA seem right to me, but it will be a bit "wait and see".

I tested SciPy it has a fair bit of failures that seem managable.

*Overall, I am not sure about this, but need the decision whether we do this, or stick with the original gh-21875 proposal.  The other PR would be a bit simpler, but I am now not quite convinced that coupling the two is worth the simplification there.  My worry is that the MA code may be strange, but downstream might also have similar hard to analyze issues with the change.*

### CODE REVIEW

File: numpy/ma/core.py [hunk 1]
```diff
@@ -201,7 +201,13 @@ def _recursive_fill_value(dtype, f):
...
     """
     if dtype.names is not None:
-        vals = tuple(_recursive_fill_value(dtype[name], f) for name in dtype.names)
+        # We wrap into `array` here, which ensures we use NumPy cast rules
+        # for integer casts, this allows the use of 99999 as a fill value
+        # for int8.
+        # TODO: This is probably a mess, but should best preserve behavior?
+        vals = tuple(
+                np.array(_recursive_fill_value(dtype[name], f))
+                for name in dtype.names)
```
Comment (mattip) [inline numpy/ma/core.py:210]: Did you reach the conclusion that this is correct based on a test failure? I don't see an added test, but the lines are not marked by coverage so I assume they are tested.

File: numpy/core/src/umath/ufunc_object.c [hunk 2]
```diff
@@ -2983,11 +2984,21 @@ PyUFunc_Reduce(PyUFuncObject *ufunc,
...
+    /*
+     * Promote and fetch ufuncimpl (currently needed to fix up the identity).
+     */
+    PyArray_Descr *descrs[3];
+    PyArrayMethodObject *ufuncimpl = reducelike_promote_and_resolve(ufunc,
+            arr, out, signature, NPY_FALSE, descrs, "reduce");
+    if (ufuncimpl == NULL) {
+        return NULL;
+    }
+
```
Comment (mattip) [inline numpy/core/src/umath/ufunc_object.c:2996]: There was a request from Numba to expose the ufunc loop used for a dtype. They did not explicitly ask for reductions, but I imagine that will be the next request. Will it be sufficient to expose `reducelike_promote_and_resolve` or will we need more logic?

File: numpy/ma/core.py (hunk 1 above)
Comment (seberg) [reply numpy/ma/core.py:210]: Yes, there were a few test failures (mainly in the recarray code that handles a lot of masked arrays with structs).
They go away with this change, which mirrors the non-struct path (that always wraps into `array()`...

It took me a while to figure that this seems like the easy way out, I don't particularly love it, but it seems to work.  It is a bit unclear if downstream might notice these changes.

**Comment by mattip** [comment]
> To me the decision here seems mainly whether we prefer this churn, over the churn of having a special (internal for now) API for Python integers

I would prefer this churn, and not have a special API for Python integers.

> I tested SciPy it has a fair bit of failures that seem managable.

Hmm, that is kind of worrisome. But doesn't #21875 also have a number of SciPy failures?

File: numpy/ma/core.py (hunk 1 above)
Comment (mattip) [reply numpy/ma/core.py:210]: +1

**Comment by seberg** [comment]
> But doesn't https://github.com/numpy/numpy/pull/21875 also have a number of SciPy failures?

Yes, but they are a fully distinct set of failures.  Merging this has no influence on those failures (besides maybe printing slightly different or giving two warnings for the same thing).

The failures here are things like `np.array([-1], dtype=np.uint8)` or `np.uint8(-1)` mainly, which NEP 50 has no opinion about.
The failures in the other PR are things like `np.uint8(3) + (-1)` which is similar but distinct (because in a pre NEP 50 world we convert to `np.int16(-1)` and not `np.uint8(-1)`.

File: numpy/core/src/multiarray/arraytypes.c.src [hunk 3]
```diff
@@ -217,6 +235,85 @@ MyPyLong_AsUnsigned@Type@ (PyObject *obj)
...
+             * (legacy promotion will prevent overflows by promoting up)
+             * so a normal deprecation makes sense.
+             * When weak promotion is active, we use "future" behavior unless
+             * warnings were explicitly opt-in.
+             */
+            if (PyErr_WarnFormat(PyExc_DeprecationWarning, 1,
+                    "NumPy will stop allowing conversion of out-of-bound "
+                    "Python integers to integer arrays.  The conversion "
+                    "of %.100R to %S will fail in the future.",
+                    obj, descr) < 0) {
```
Comment (mattip) [inline numpy/core/src/multiarray/arraytypes.c.src:295]: Maybe add the `astype()` workaround, i.e. `np.array(-1, np.uint8)` will raise here, but `np.array(-1).astype(np.uint8)` will do the right thing? I am not sure how to express that this deep in the code path...

Comment (seberg) [reply numpy/core/src/multiarray/arraytypes.c.src:298]: Good idea, also added a release note.  If this becomes too painful for downstream, we could consider allowing `np.uint8(-1)` making it explicitly work for negative values.  (I would still restrict this to the range of the corresponding signed integer, so `-128` for `uint8`)

Comment (mattip) [reply numpy/core/src/multiarray/arraytypes.c.src:298]: we might need that as a fallback

File: doc/release/upcoming_changes/22393.deprecation.rst [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+------------------------------------------
+Attempting a conversion from a Python integer to a NumPy
+value will now always check whether the result can be
+represented by NumPy.  This means the following examples will
+fail in the future and give a ``DeprecationWarning`` now::
+
+    np.uint8(-1)
+    np.array([3000], dtype=np.int8)
+
+While many of these did succeed before.  Such code was mainly
```
Comment (mattip) [inline doc/release/upcoming_changes/22393.deprecation.rst:11]:
```suggestion
Many of these did succeed before.  Such code was mainly
```

**Review by mattip — APPROVED** [review]
LGTM. Only a small nit in the release note. Anyone else want to take a look?

**Comment by mattip** [comment]
Thanks @seberg. What happens now to #21875?

**Comment by seberg** [comment]
I will update it to remove the special parsing path.  They will use the normal parsing paths instead always (might need some test adjustments).  It won't be much simpler, but we won't have the "special" path at least.

**Comment by WarrenWeckesser** [comment]
Fixing the SciPy tests wasn't too bad: https://github.com/scipy/scipy/pull/17209

**Pull request merged**